app/src/main/java/tech/relaycorp/courier/ui/settings/SettingsViewModel.kt

```diff
@@ -17,28 +23,85 @@ import javax.inject.Inject
 class SettingsViewModel
 @Inject constructor(
     observeStorageUsage: ObserveStorageUsage,
-    deleteAllStorage: DeleteAllStorage
+    deleteAllStorage: DeleteAllStorage,
+    storagePreferences: StoragePreferences,
+    diskStats: DiskStats
 ) : BaseViewModel() {
 
     // Inputs
 
     private val deleteDataClicks = PublishChannel<Click>()
     fun deleteDataClicked() = deleteDataClicks.sendBlocking(Click)
 
+    private val maxStorageChanged = PublishChannel<StorageSize>()
+    fun maxStorageChanged(value: StorageSize) = maxStorageChanged.sendBlocking(value)
+
     // Outputs
 
     private val deleteDataEnabled = BehaviorChannel<EnableState>()
     fun deleteDataEnabled() = deleteDataEnabled.asFlow()
 
+    private val maxStorage = BehaviorChannel<StorageSize>()
+    fun maxStorage() = maxStorage.asFlow()
+
+    private val maxStorageBoundary = BehaviorChannel<Boundary<StorageSize>>()
+    fun maxStorageBoundary() = maxStorageBoundary.asFlow()
+
+    private val storageStats = BehaviorChannel<StorageStats>()
+    fun storageStats() = storageStats.asFlow()
+
     init {
         observeStorageUsage
             .observe()
             .onEach { deleteDataEnabled.send((!it.usedByApp.isZero).toEnableState()) }
             .launchIn(ioScope)
 
+        combine(
+            observeStorageUsage.observe(),
+            diskStats.observeAvailableStorage()
+        ) { usage, available ->
+            StorageStats(
+                used = usage.usedByApp,
+                usedPercentage = usage.percentage,
+                available = available,
+                total = diskStats.getTotalStorage()
+            )
+        }
+            .onEach(storageStats::send)
+            .launchIn(ioScope)
+
         deleteDataClicks
             .asFlow()
             .onEach { deleteAllStorage.delete() }
             .launchIn(ioScope)
+
+        storagePreferences
+            .getMaxStorageSize()
+            .onEach(maxStorage::send)
+            .launchIn(ioScope)
+
+        maxStorageChanged
+            .asFlow()
+            .onEach { storagePreferences.setMaxStorageSize(it) }
+            .launchIn(ioScope)
+
+        ioScope.launch {
+            val totalStorageValue = diskStats.getTotalStorage()
+            maxStorageBoundary.send(
+                Boundary(
+                    MIN_STORAGE_SIZE,
+                    totalStorageValue,
+                    STORAGE_SIZE_STEP
+                )
+            )
+        }
+    }
+
+    companion object {
+        @VisibleForTesting
+        val MIN_STORAGE_SIZE = StorageSize(500_000_000)
+
+        @VisibleForTesting
+        val STORAGE_SIZE_STEP = StorageSize(500_000_000)
```

Comment:
Shouldn't the step be much lower? Or are these just some dummy values?

Reply:
It's up for discussion, how many steps we want to have. Also, it will depend if we keep the seek bar. A seek bar handles it well, a dropdown won't :)

Reply:
Agreed, but there are two more things to factor in: What's the mean disk space in the regions we're targeting?  Would the user feel starting at half a gigabyte is excessive?

I don't know the answers to those questions though. I tried to find stats on the former but couldn't find any. And I don't think we'll find an objective answer to the latter unless we do some user testing, which would be though right now (although we can certainly flag this now, so it's considered in the UX assessment OTF will be commissioning).

I think these issues would go away if we made these values relative to the total internal disk space. Like the max value. Thoughts?

Reply:
The minimum value we can basically set to how little do you need for the Courier app to operate normally? At least some 50 or 100MB? Then we can step based on that.

Reply:
Agreed.

A cargo message can take up anywhere from 6 kib to [8 MiB](https://github.com/relaynet/specs/issues/21). (The lower bound will drop massively once we implement https://github.com/relaynet/specs/issues/21).)

Which means:

- With 50 MB, you could store between 6 and 8k cargoes.
- With 100 MB, you could store between 12 and 16k cargoes.

So I'd suggest 100 MB as the minimum. Wdyt?

Reply:
Yup, let's go with 100MB minimum and a 100MB step.